src/prefetch.cpp

```diff
@@ -0,0 +1,325 @@
+        }
+
+        unlink(m_lock_file_path.c_str());
+
+        int ok_fd = open(m_ok_file_path.c_str(), O_RDWR | O_CREAT | O_TRUNC | O_EXCL, 0666);
+        if (ok_fd < 0) {
+            LOG_ERRNO_RETURN(0, -1, "Prefetch: open OK file failed");
+        }
+        close(ok_fd);
+        return 0;
+    }
+
+    int reload(size_t trace_file_size) {
+        // Reload header
+        TraceHeader hdr = {};
+        ssize_t n_read = m_trace_file->read(&hdr, sizeof(TraceHeader));
+        if (n_read != sizeof(TraceHeader)) {
+            LOG_ERRNO_RETURN(0, -1, "Prefetch: reload header failed");
+        }
+        if (TRACE_MAGIC != hdr.magic) {
+            LOG_ERROR_RETURN(0, -1, "Prefetch: trace magic mismatch");
+        }
+        if (trace_file_size != hdr.data_size + sizeof(TraceHeader)) {
+            LOG_ERROR_RETURN(0, -1, "Prefetch: trace file size mismatch");
+        }
+
+        // Reload content
+        uint32_t checksum = 0;
+        TraceFormat fmt = {};
+        for (int i = 0; i < hdr.data_size / sizeof(TraceFormat); ++i) {
+            n_read = m_trace_file->read(&fmt, sizeof(TraceFormat));
+            if (n_read != sizeof(TraceFormat)) {
+                LOG_ERRNO_RETURN(0, -1, "Prefetch: reload content failed");
+            }
+            checksum = crc32::crc32c_extend(&fmt, sizeof(TraceFormat), checksum);
+            // Save in memory
+            m_replay_queue.push(fmt);
+        }
+
+        if (checksum != hdr.checksum) {
+            queue<TraceFormat> tmp;
+            m_replay_queue.swap(tmp);
+            LOG_ERROR_RETURN(0, -1, "Prefetch: reload checksum error");
+        }
+
+        LOG_INFO("Prefetch: Reload ` records", m_replay_queue.size());
+        return 0;
+    }
+
+    int detect_lock() {
+        while (true) {
+            photon::thread_sleep(1);
+            if (access(m_lock_file_path.c_str(), F_OK) != 0) {
+                m_record_stopped = true;
+                dump();
+                break;
+            }
+        }
+        return 0;
+    }
+
+    friend LogBuffer& operator<<(LogBuffer& log, const PrefetcherImpl::TraceFormat& f);
+};
+
+LogBuffer& operator<<(LogBuffer& log, const PrefetcherImpl::TraceFormat& f) {
+    return log << "Op " << char(f.op) << ", Count " << f.count << ", Offset " << f.offset << ", Layer_index "
+               << f.layer_index;
+}
+
+PrefetchFile::PrefetchFile(IFile* src_file, uint32_t layer_index, Prefetcher* prefetcher) :
+        ForwardFile(src_file),
+        m_layer_index(layer_index),
+        m_prefetcher((PrefetcherImpl*) prefetcher) {
+    if (m_prefetcher->get_mode() == PrefetcherImpl::Mode::Replay) {
+        m_prefetcher->register_src_file(layer_index, src_file);
+    }
+}
+
+ssize_t PrefetchFile::pread(void* buf, size_t count, off_t offset) {
+    if (m_prefetcher->get_mode() == PrefetcherImpl::Mode::Record) {
+        m_prefetcher->record(PrefetcherImpl::TraceOp::READ, m_layer_index, count, offset);
```

Comment:
`record` happend before actual `pread` activity, ignored the result of `pread` action.
Is this record properly in case of `pread` failure, or read till EOF?

Reply:
record only successful `pread` action might be more effective

Reply:
OK